### PR TITLE
Firefly 1050 Converting class based to functional components

### PR DESCRIPTION
**Pull request by kpuriIpac**
#### [Firefly-1050](https://jira.ipac.caltech.edu/browse/FIREFLY-1050): converting 3 class components to functional components
 - Converted the Menu, TimePanelView and DateTimePicker class components to functional components 
 - Made a small change to menu.jsx's makeMenuItem function arguments (set default values for isWorking and badgeCount) 

#### Testing
 -  https://fireflydev.ipac.caltech.edu/firefly-1050-class-to-function/firefly/
 - go to TAP -> service: CADC -> table collection: ivoa -> table: ivoa.ObsFile or ivoa.ObsPart 
 - look at the UI for time field entry and date time picker (click on the calendar icon)
 - Functionality should be the same as before, as the only changes made were class components -> functional components

### CODE REVIEW

**Review by loitly — APPROVED** [review]
For the requested task, changes look good.  Ok to merge.

File: src/firefly/js/ui/TimePanel.jsx [hunk 1]
```diff
@@ -19,72 +19,64 @@ const iconMap = {'calendar': {icon: CALENDAR, title: 'Show the calendar for sele
...
+
+    const outsideWidth = inputWidth + 6;
+    const timePart = iconField ? (<div style={{position: 'relative', width: outsideWidth}}>
+                                        {timeField}
+                                        {iconField}
+                                  </div>)
+                                : (timeField);
+
+    const newFeedbackStyle = clone(feedbackStyle, {width: inputWidth});
+    const lStyle = clone(labelStyle, {whiteSpace:'nowrap'});
```
Comment (robyww) [inline src/firefly/js/ui/TimePanel.jsx:64]: We are getting away from using `clone` or `Object.assign`.  These two lines should be:
```javascript
 const newFeedbackStyle = {...feedbackStyle, width: inputWidth};
 const lStyle = {...labelStyle, whiteSpace:'nowrap'};
```

File: src/firefly/js/ui/TimePanel.jsx [hunk 2]
```diff
@@ -19,72 +19,64 @@ const iconMap = {'calendar': {icon: CALENDAR, title: 'Show the calendar for sele
...
+    const labelDiv = (<div style={lStyle}>{label}</div>);
+    const timeDiv = (
+        <div>
+            {timePart}
+            <TimeFeedback {...{showHelp, feedback, feedbackStyle: newFeedbackStyle, examples, timeMode}}/>
```
Comment (robyww) [inline src/firefly/js/ui/TimePanel.jsx:69]: same type of syntax I mentioned above, make sure you understand it.

File: src/firefly/js/ui/TimePanel.jsx [hunk 3]
```diff
@@ -19,72 +19,64 @@ const iconMap = {'calendar': {icon: CALENDAR, title: 'Show the calendar for sele
...
+        </div>
+    );
+    const flexDirection = (labelPosition && labelPosition === 'top') ? 'column' : 'row';
```
Comment (robyww) [inline src/firefly/js/ui/TimePanel.jsx:72]: simplify to:
```javascript
const flexDirection = (labelPosition === 'top') ? 'column' : 'row';
```

File: src/firefly/js/ui/DateTimePickerField.jsx [hunk 4]
```diff
@@ -80,94 +81,76 @@ function isSameTimes(time1, time2) {
      return compareMoments(m1, m2);
 }
 
-export class DateTimePicker extends PureComponent {
-    constructor(props) {
-        super(props);
-        this.onClose = this.onClose.bind(this);
-        this.onSelectedDate = this.onSelectedDate.bind(this);
+export function DateTimePicker({showInput,openPicker, value='', onChange, onChangeOpenStatus,wrapperStyle,
+                                   inputStyle, groupKey, timeFieldKey}) {
```
Comment (robyww) [inline src/firefly/js/ui/DateTimePickerField.jsx:85]: groupKey should be added to the propTypes (eslint caught this)

File: src/firefly/js/ui/DateTimePickerField.jsx [hunk 5]
```diff
@@ -80,94 +81,76 @@ function isSameTimes(time1, time2) {
...
 
-        const {value = ''} = props;
-        const timeMoment = aMoment(value).isValid() ? aMoment(value) : value;   // a moment or original string value
-        this.state={timeMoment};  // store valid moment or string
-    }
-
-    onClose() {
-        const {onChangeOpenStatus} = this.props;
+    const [timeMoment, setTimeMoment] = useState(aMoment(value).isValid() ? aMoment(value) : value);
```
Comment (robyww) [inline src/firefly/js/ui/DateTimePickerField.jsx:87]: Convert this to a function:

```javascript
const [timeMoment, setTimeMoment] = useState(() => aMoment(value).isValid() ? aMoment(value) : value);
```
This reason is that react will only call the function the first time it renders. If you don't make a function than the work is done every time even though react only uses it the first time.
Any computation in useState initialization should be a function.

File: src/firefly/js/ui/DateTimePickerField.jsx [hunk 6]
```diff
@@ -80,94 +81,76 @@ function isSameTimes(time1, time2) {
...
         const newTime = tryConvertToMoment(moment, true);
         if (typeof moment === 'string' && !newTime.isValid()) {
             moment = '';
         }
 
-        this.setState({timeMoment: moment});
+        setTimeMoment(moment);
         if (onChange) {
             onChange(moment);
         }
```
Comment (robyww) [inline src/firefly/js/ui/DateTimePickerField.jsx:127]: Simplify to:
```javascript
onChange?.(moment)
```

File: src/firefly/js/ui/DateTimePickerField.jsx [hunk 7]
```diff
@@ -80,94 +81,76 @@ function isSameTimes(time1, time2) {
...
+                }
+            });
+        }
 
+        //componentWillUnmount code
+        return () => {
+            if (unbinder) unbinder();
+            iAmMounted = false;
+        };
+    }, []);
```
Comment (robyww) [inline src/firefly/js/ui/DateTimePickerField.jsx:109]: I told you I would suggest another way to do this in the code review. Here it is. using our recently written `useFieldGroupValue` (it is kind of like `useState`)

```javascript
   const [setTimeStr,getTimeStr]= useFieldGroupValue(timeFieldKey);

   useEffect(()  => {
         const time= getTimeStr();
         if (!time || !isSameTimes(time, timeMoment)) return;
        setTimeMoment(time);
  }, [getTimeStr]);
```

If you are having trouble make this work. Then we can look at it together. The hook is pretty new so there might be some issues.

**Review by robyww — COMMENTED** [review]
Your code looks good. Most of my comments we just updating existing code to more modern JavaScript. We try to do that when we update components like this.

Testing all looked good.

A couple of other things.

- Make sure your ESlint is catching unused imports. Pay attention to anything is highlights.
- Try out my suggestion for the our new hook `useFieldGroupValue`

File: src/firefly/js/ui/DateTimePickerField.jsx (hunk 7 above)
Comment (kpuriIpac) [reply src/firefly/js/ui/DateTimePickerField.jsx:109]: Hi Trey, should the if condition above be this instead (removed the 'not' from isSameTimes): 
 
```javascript
         if (!time || isSameTimes(time, timeMoment)) return;
```

to retain the logic of the original code? We only call setTimeMoment(time) when time != timeMoment, correct?

Comment (robyww) [reply src/firefly/js/ui/DateTimePickerField.jsx:109]: opps. good catch. yes it should have been
```
if (!time || !isSameTimes(time, timeMoment)) return;
```

File: src/firefly/js/ui/DateTimePickerField.jsx [hunk 8]
```diff
@@ -8,6 +8,7 @@ import FieldGroupUtils from '../fieldGroup/FieldGroupUtils';
 
 import './tap/react-datetime.css';
 import {useFieldGroupConnector} from './FieldGroupConnector.jsx';
+import {ISO} from "firefly/ui/tap/TapUtil";
```
[the remark this answers is not in the document]
Comment (kpuriIpac) [reply src/firefly/js/ui/DateTimePickerField.jsx:11]: Yes, not sure why I'd added this. Removed now.

**Pull request merged**